### ISSUE TITLE
[C#] Enum inspector values are incorrectly determined when serializing

### ISSUE THREAD

**Issue by mattbenic** [AtomicGameEngine/AtomicGameEngine#1534]
Currently, it appears that enum values set in the inspector are set by assigning the value corresponding to an enum's position in the list of enum values, rather than the actual enum value. For example, TextureUnit is defined (in C#) as:
```
   public enum TextureUnit
   {
      TU_DIFFUSE = 0,
      TU_ALBEDOBUFFER = 0,
      TU_NORMAL = 1,
      TU_NORMALBUFFER = 1,
      TU_SPECULAR = 2,
      TU_EMISSIVE = 3,
      TU_ENVIRONMENT = 4,
      TU_LIGHTRAMP = 5,
      TU_LIGHTSHAPE = 6,
      TU_SHADOWMAP = 7,
      MAX_MATERIAL_TEXTURE_UNITS = 5,
      MAX_TEXTURE_UNITS = 8
   }
```

If a wrapper enum like the following is used (to get around #990):
```
public enum TextureUnit
{
    TU_DIFFUSE = AtomicEngine.TextureUnit.TU_DIFFUSE,
    TU_ALBEDOBUFFER = AtomicEngine.TextureUnit.TU_ALBEDOBUFFER,
    TU_NORMAL = AtomicEngine.TextureUnit.TU_NORMAL,
    TU_NORMALBUFFER = AtomicEngine.TextureUnit.TU_NORMALBUFFER,
    TU_SPECULAR = AtomicEngine.TextureUnit.TU_SPECULAR,
    TU_EMISSIVE = AtomicEngine.TextureUnit.TU_EMISSIVE,
    TU_ENVIRONMENT = AtomicEngine.TextureUnit.TU_ENVIRONMENT,
    MAX_MATERIAL_TEXTURE_UNITS = AtomicEngine.TextureUnit.MAX_MATERIAL_TEXTURE_UNITS,
    TU_LIGHTSHAPE = AtomicEngine.TextureUnit.TU_LIGHTSHAPE,
    TU_SHADOWMAP = AtomicEngine.TextureUnit.TU_SHADOWMAP,
    MAX_TEXTURE_UNITS = AtomicEngine.TextureUnit.MAX_TEXTURE_UNITS
}

```
And a TextureUnit field exposed to the inspector:
```
    [Inspector]
    private TextureUnit textureUnit = TextureUnit.TU_DIFFUSE;

```

Setting the value to TU_EMISSIVE (3) in the editor will result in the following saved in the scene file:
`<variant name="textureUnit" type="Int" value="5" />`

**Comment by JoshEngebretson** [comment]
The texture unit enum is a special case (with special handling) as it has preprocessor guards/values for desktop/mobile: https://github.com/AtomicGameEngine/AtomicGameEngine/blob/1574c8d0803b384b6045e00e1a73b13426ec7afe/Source/Atomic/Graphics/GraphicsDefs.h#L326

I will look into this today to see if it is a general issue or with this enum specifically

**Comment by JoshEngebretson** [comment]
This turned out to be a bit more involved than I first thought,  part of which was a misguided effort to make it possible to set enum fields both by numeric value and string enum name... 

I have a branch up here: https://github.com/AtomicGameEngine/AtomicGameEngine/tree/JME-ATOMIC-1534

... and a test here: https://github.com/AtomicGameEngine/AtomicTests/tree/JME-ATOMIC-ENUMUPDATE/CSharpTests/InspectorEnumTest

A groovy thing is that the script fields now serialize via the *name* of the enum value instead of the numeric value, which makes them much more robust for changes to the enum (*and is how they should have been done initially for script fields, oops*)

```xml
<attribute name="FieldValues">
    <variant name="speed" type="Float" value="2" />
    <variant name="textureUnit" type="String" value="TU_EMISSIVE" />
</attribute>
```

I want to get a bit more testing into this and add some comments before landing, which I plan for tomorrow.

**Comment by mattbenic** [comment]
Persisting them by enum name does make more sense. It's also quite possible that the enum names might change, so as far as change resilience goes you just can't win ;)

**Comment by JoshEngebretson** [comment]
Yeah, in he case of the name changing the best we can do is change to default, which is what happens now.

This was a short odyssey, check out #1538

I made CSComponentAssembly type information available at runtime, which was necessary to map the enum strings to corresponding values.  It is also something that will be useful to have moving forward, and really can't be generated on the fly (some of it requires deep assembly introspection that isn't available at runtime).

The Cache file is generated when the assembly is inspected by the AtomicNETService, if that process is not available for whatever reason (like running from AtomicTool directly), then the CSComponentAssembly cached type info in the .asset file will be used when generating the Cache file.

I did build out an AtomicTest (https://github.com/AtomicGameEngine/AtomicTests/tree/master/CSharpTests/InspectorEnumTest)  and have tested deploying to Windows and Android with the new improved enums, let me know if you run into any problems.  The cache representation of the component assemblies is simply JSON located in Cache/```guid```, so whatever the assembly GUID is in the .asset file.  It would be good to verify that this is being generated properly locally and on CI for the project.
